### PR TITLE
Various performance improvements

### PR DESCRIPTION
**Pull request by WhoIsJohannes**
Please see comments on the individual commits. I tried to explain everything there with screenshots from the profiler.

### CODE REVIEW

File: core/src/com/unciv/logic/map/tile/Tile.kt [hunk 1]
```diff
@@ -232,7 +232,7 @@ open class Tile : IsPartOfGameInfoSerialization {
     }
 
     fun isExplored(player: Civilization): Boolean {
-        if (DebugUtils.VISIBLE_MAP || player.isSpectator())
+        if (DebugUtils.VISIBLE_MAP || player.civName == Constants.spectator)
```
Comment (yairm210) [inline core/src/com/unciv/logic/map/tile/Tile.kt:235]: This is actually worse performance wise, no?

**Comment by WhoIsJohannes** [comment]
It's called like 10 million times just in one next turn evaluation for me.
I also suspect the lazy and just objects being dereferenced / passed around.

Il gio 27 apr 2023, 17:50 SomeTroglodyte ***@***.***> ha
scritto:

**Comment by yairm210** [comment]
@SomeTroglodyte This is still a much smaller performance hit than string comparisons - that's why those lazies exist in the first place, as a result of performance debugging (#9098)

Doing `val x = y` won't work with json-unpacked classes, since they are initially empty and are only filled on json loading.
Do you have any suggestion for an alternative?

**Comment by WhoIsJohannes** [comment]
Thanks for all the brain power put into this. Have you guys seen my
screenshots? It's definitely faster this way, believe it or not 😋.
Unfortunately they got lost a bit, only through clicking on some comment
icon somewhere will bring them up since I put them before sending the PR.

Il sab 29 apr 2023, 21:30 SomeTroglodyte ***@***.***> ha
scritto:

**Comment by yairm210** [comment]
There's actually a simple solution for this use case - since the isSpectator is in the *civ* class and not the *nation* class, we can use a var bool in the civ class, and set it in setTransients, this avoiding the problem entirely :D

**Comment by yairm210** [comment]
This is another one of those 'too much going on' PRs, if you split it into different PRs then we can merge the simple fixes and argue about the contentious ones :)

**Comment by yairm210** [comment]
Relevant to the lazy evaluation (search for "negative latency"):
https://apenwarr.ca/log/20230415



Tldr of the rest of the post:
- The hard part of science is generating testable hypotheses, not the actual testing
- Understanding the underlying components of a system *does not* mean you understand the system
- Engineering is "getting stuff done passably", not perfectly

- To get things done efficiently, to do real engineering, sometimes you need to *guess stuff* (hypothesis) and then test it, and that requires *insight* or system design understanding. The open question is, will AIs be able to do that

**Comment by WhoIsJohannes** [comment]
How about this: I revert the lazy part and we figure it out separately. Is
that the only part that we're discussing at this point?

Il dom 30 apr 2023, 17:34 SomeTroglodyte ***@***.***> ha
scritto:

**Comment by yairm210** [comment]
I'm also not 100% sure that the movement improvements don't change anything but there's so much going on here that I find it hard to review

**Comment by yairm210** [comment]
I am willing to go the other way though - let's merge this and then see what happens/if we can improve on it 😜

**Comment by WhoIsJohannes** [comment]
Thanks! I was hoping the individual commits would help, but I get it. I'm
always a bit worried building PRs on top of each other because I don't want
to get into merging hell... And I somehow need to run everything together
locally. Anyways, for now I'm done with performance I think and I'm not
planning to do much any time soon. We're travelling for a long time and
I'll have to start my real job after that. Also there isn't so much right
now I really think needs doing (other than AI, but that is not something I
could do quickly).

Il lun 1 mag 2023, 06:35 Yair Morgenstern ***@***.***> ha
scritto:

**Comment by yairm210** [comment]
Good luck!
